helix-contract/contracts/ln/base/LnBridgeTarget.sol

```diff
@@ -0,0 +1,130 @@
+        uint48 nonce;
+        uint48 lastRefundNonce;
+        address slasher;
+    }
+
+    mapping(bytes32 => TransferInfo) public transferInfos;
+    mapping(bytes32 => address) public refundReceiver;
+
+    event TransferFilled(bytes32 transferId, address slasher);
+
+    // fill transfer
+    // 1. if transfer is not refund or relayed, LnProvider relay message to fill the transfer, and the transfer finished on target chain
+    // 2. if transfer is timeout and not processed, slasher(any account) can fill the transfer and request refund
+    // if it's filled by slasher, we store the address of the slasher
+    function _fillTransfer(
+        TransferParameter calldata params,
+        address slasher
+    ) internal returns(bytes32 transferId, uint48 lastRefundNonce) {
+        TransferInfo memory lastInfo = transferInfos[params.lastTransferId];
+        require(lastInfo.nonce + 1 == params.nonce, "Invalid last transferId");
+        transferId = keccak256(abi.encodePacked(
+            params.lastTransferId,
+            params.lastBlockHash,
+            params.nonce,
+            params.timestamp,
+            params.token,
+            params.receiver,
+            params.amount));
+        TransferInfo memory transferInfo = transferInfos[transferId];
+        require(transferInfo.nonce == 0, "lnBridgeTarget:message exist");
+        lastRefundNonce = lastInfo.slasher != address(0) ? params.nonce - 1 : lastInfo.lastRefundNonce;
+        transferInfos[transferId] = TransferInfo(params.nonce, lastRefundNonce, slasher);
+        if (params.token == address(0)) {
+            require(msg.value >= params.amount, "lnBridgeTarget:invalid amount");
+            payable(params.receiver).transfer(params.amount);
+        } else {
+            _safeTransferFrom(params.token, msg.sender, params.receiver, uint256(params.amount));
+        }
+        emit TransferFilled(transferId, slasher);
+    }
+
+    function transferAndReleaseMargin(TransferParameter calldata params) payable external {
+        _fillTransfer(params, address(0));
+    }
+
+    // The condition for slash is that the transfer has timed out
+    // Meanwhile we need to request a refund transaction to the source chain to withdraw the LnProvider's margin
+    // On the source chain, we need to verify all the transfers before has been relayed or slashed.
+    // So we needs to carry the the previous refund transferId to ensure that the slash is continuous.
+    function _slashAndRemoteRefund(
+        TransferParameter calldata params,
+        bytes32 lastRefundTransferId
+    ) internal returns(bytes memory message) {
+        require(block.timestamp > params.timestamp + MIN_REFUND_TIMESTAMP, "refund time not expired");
+        (bytes32 transferId, uint48 lastRefundNonce) = _fillTransfer(params, msg.sender);
+        TransferInfo memory lastRefundInfo = transferInfos[lastRefundTransferId];
+        require(lastRefundInfo.nonce == lastRefundNonce, "invalid last refund nonce");
```

Comment:
```suggestion
        // The same nonce indicate they have same transfer Id.
        require(transferInfos[lastRefundTransferId].nonce == transferInfos[transferId].nonce, "invalid last refund nonce");
```

>  // The same nonce indicate they have same transfer Id.
Is this correct? nonce can be fake? Which could cause `lastRefundTransferId ` to be non exists, is it ok for source chain?